### PR TITLE
Generic column operators and update Index signature.

### LINKED ISSUE #82: Datetime desc in Index results in arg-type error
**Describe the bug**
mypy reports an error when `datetime_col.desc()` is used as an `Index` second argument; `datetime_col` (without `desc`) works as intended.

**Expected behavior**
mypy does not produce an error

**To Reproduce**
```py
from sqlalchemy import Column, DateTime, Integer, Index
from sqlalchemy.ext.declarative import declarative_base

Base = declarative_base()


class Model(Base):
    
    id = Column(Integer, primary_key=True)
    created_at = Column(DateTime)

    __tablename__ = 'model'
    __table_args__ = (
        # error: Argument 2 to "Index" has incompatible type "ClauseElement"; expected "ColumnClause[Any]"  [arg-type]
        Index('created_at_idx', created_at.desc()),
    )
```

**Error**

```
error: Argument 2 to "Index" has incompatible type "ClauseElement"; expected "ColumnClause[Any]"  [arg-type]
```

**Versions.**
 - OS: dockerized python:3.9.4-buster
 - Python: 3.9.4
 - SQLAlchemy: 1.4.12

**Have a nice day!**

### PR DESCRIPTION
**Pull request by MaicoTimmerman**
### Description
Make `ColumnOperators` generic, needed to update `Index.__init__`.

Fixes #82 

### Checklist

This pull request is:

- [ ] A documentation / typographical error fix
	- Good to go, no issue or tests are needed
- [x] A short code fix
	- please include the issue number, and create an issue if none exists, which
	  must include a complete example of the issue.  one line code fixes without an
	  issue and demonstration will not be accepted.
	- Please include: `Fixes: #<issue number>` in the commit message
	- please include tests.   one line code fixes without tests will not be accepted.
- [ ] A new feature implementation
	- please include the issue number, and create an issue if none exists, which must
	  include a complete example of how the feature would look.
	- Please include: `Fixes: #<issue number>` in the commit message
	- please include tests.

**Have a nice day!**

### CODE REVIEW

File: sqlalchemy-stubs/sql/schema.pyi [hunk 1]
```diff
@@ -545,7 +545,11 @@ class Index(DialectKWArgs, ColumnCollectionMixin, SchemaItem):
     unique: bool = ...
     expressions: List[Column[Any]] = ...
     def __init__(
-        self, name: str, *expressions: ColumnClause[Any], **kw: Any
+        self,
+        name: str,
+        # TODO This is technically DDLConstraintColumnRole, do we use roles for types?
+        *expressions: Union[ColumnClause[Any], ColumnElement[Any], TextClause],
```
Comment (bryanforbes) [inline sqlalchemy-stubs/sql/schema.pyi:551]: From my conversation with @CaselIT and @zzzeek, we can't use roles for types. However, I don't remember why off the top of my head.

**Review by bryanforbes — CHANGES_REQUESTED** [review]

File: sqlalchemy-stubs/sql/schema.pyi (hunk 1 above)
Comment (CaselIT) [reply sqlalchemy-stubs/sql/schema.pyi:551]: I don't actually remember.

File: sqlalchemy-stubs/sql/operators.pyi [hunk 2]
```diff
@@ -77,137 +83,122 @@ class custom_op:
...
 
 def commutative_op(fn: _F) -> _F: ...
 def comparison_op(fn: _F) -> _F: ...
-def from_() -> None: ...
-def function_as_comparison_op() -> None: ...
-def as_() -> None: ...
-def exists() -> None: ...
-def is_true(a: Any) -> None: ...
-
-istrue = is_true
```
[the remark this answers is not in the document]
Comment (MaicoTimmerman) [reply sqlalchemy-stubs/sql/operators.pyi:165]: The `istrue`, `isfalse`, `isnot_distinct_from`, `isnot`, `notlike_op`, `notilike_op`, `notbetween_op`, and `notin_op` are deprecated https://github.com/sqlalchemy/sqlalchemy/blob/master/lib/sqlalchemy/sql/operators.py#L1312.

I felt that there is no use keeping them in the stubs, as people should be using the functions on the RHS of these assignments anyway. I can return them in the stubs, with added similar comments as in the original source.

[the remark this answers is not in the document]
Comment (bryanforbes) [reply sqlalchemy-stubs/sql/operators.pyi:165]: I would add them to the end of the file with a note that they're deprecated

File: sqlalchemy-stubs/sql/schema.pyi (hunk 1 above)
Comment (zzzeek) [reply sqlalchemy-stubs/sql/schema.pyi:551]: it's not typically very applicable to schema things like Index, but to support ORM use cases, there's the single case of an ORM mapped class that is user defined and therefore does not suit any kind of role directly, it has to go through inspect() first:



    class MyClass(Base):
         # ...

    stmt = select(MyClass)    # <---- no role

that's pretty much the only real world use case right now.

Comment (zzzeek) [reply sqlalchemy-stubs/sql/schema.pyi:551]: OK so adding more completeness, anything that either responds to inspect(), or anything that has a `__clause_element__()` method, which we call to get the thing.  However, real-world use of `__clause_element__()` is always on an object that suits a role in any case.

Comment (zzzeek) [reply sqlalchemy-stubs/sql/schema.pyi:551]: I would say for now that anything that accepts column elements / text like Index above and does not accept FROM objects is probably fine to use roles.    Really the thing with ORM classes being sent is `select()`, `select().select_from()`, and ...maybe really not much more than that actually.

Comment (CaselIT) [reply sqlalchemy-stubs/sql/schema.pyi:551]: `join` also comes to mind

Comment (zzzeek) [reply sqlalchemy-stubs/sql/schema.pyi:551]: there is a separate ORMJoin construct, though I guess sqlalchemy.sql.join(ClassA, ClassB) will now do something (I dont think it did in 1.3 but I haven't tried).

Comment (CaselIT) [reply sqlalchemy-stubs/sql/schema.pyi:551]: also `select(ClassA).join(ClassB)`. I don't remember also, I used it also on 1.3, but I don't remember if it was with `ClassFoo` or `ClassFoo.__table__`

Comment (zzzeek) [reply sqlalchemy-stubs/sql/schema.pyi:551]: basically things that accept FromClause types of roles will also need to accept ORM mapped classes.  otheriwse I think roles usually work.

File: sqlalchemy-stubs/sql/operators.pyi (hunk 2 above)
[the remark this answers is not in the document]
Comment (MaicoTimmerman) [reply sqlalchemy-stubs/sql/operators.pyi:165]: Done

File: sqlalchemy-stubs/sql/schema.pyi (hunk 1 above)
Comment (MaicoTimmerman) [reply sqlalchemy-stubs/sql/schema.pyi]: I've adapted this to the correct role.

**Review by CaselIT — APPROVED** [review]
sorry for the delay 😶

**Pull request merged**